### PR TITLE
server/{market,asset}: token registration and balance checks

### PR DESCRIPTION
**Pull request by buck54321**
```
Add RegisterToken method for token registration. The asset package
keeps a map of tokens as well as the token "fee families", e.g. all
tokens of a base chain asset.

DEXBalancer.CheckBalance reworked to account for fee-family assets.
```

### CODE REVIEW

**Review by JoeGruffins — APPROVED** [review]
Lots of calculations for the balance check.

File: server/market/balancer.go [hunk 1]
```diff
@@ -77,43 +139,123 @@ func (b *DEXBalancer) CheckBalance(acctAddr string, assetID uint32, qty, lots ui
...
+			return false
+		}
+		if feeBal == 0 {
+			log.Tracef("(*DEXBalancer).CheckBalance(%q, %d, %d, %d, %d) false for zero fee asset (%s) balance",
+				acctAddr, assetID, qty, lots, redeems, feeBalancer.assetInfo.Symbol)
+			return false
+		}
+	}
+
+	var swapFees, redeemFees uint64
```
Comment (martonp) [inline server/market/balancer.go:178]: No need to track them separately since they are always summed in the end.

**Review by martonp — APPROVED** [review]
The logic checks out, I don't see any issues.

File: server/asset/driver.go [hunk 2]
```diff
@@ -75,41 +97,83 @@ func NewAddresser(name string, acctXPub string, keyIndexer KeyIndexer, network d
...
 	return drv.Setup(configPath, logger, network)
 }
 
 // Version retrieves the version of the named asset's Backend implementation.
-func Version(name string) (uint32, error) {
-	driversMtx.Lock()
-	drv, ok := drivers[name]
-	driversMtx.Unlock()
+func Version(assetID uint32) (uint32, error) {
+	drv, ok := baseDriver(assetID)
```
[the remark this answers is not in the document]
Comment (martonp) [reply server/asset/driver.go:155]: `driversMtx` is held inside `baseDriver`.

File: server/asset/driver.go [hunk 3]
```diff
@@ -11,20 +11,20 @@ import (
 )
 
 var (
-	driversMtx sync.Mutex
-	drivers    = make(map[string]Driver)
+	driversMtx  sync.Mutex // Can we get rid of this?
+	drivers     = make(map[uint32]Driver)
+	tokens      = make(map[uint32]TokenDriver)
```
[the remark this answers is not in the document]
Comment (martonp) [reply server/asset/driver.go:16]: Yes, the plan is to assign unused slip-44 IDs.

File: server/asset/driver.go [hunk 4]
```diff
@@ -75,41 +97,83 @@ func NewAddresser(name string, acctXPub string, keyIndexer KeyIndexer, network d
...
+}
+
+// Tokens returns the child tokens registered for a base chain asset.
+func Tokens(assetID uint32) map[uint32]*dex.Token {
+	m := make(map[uint32]*dex.Token, len(childTokens[assetID]))
+	for k, v := range childTokens[assetID] {
+		m[k] = v
+	}
+	return m
+}
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/asset/driver.go:179]: Except returning just `childTokens[assetID]` would be a reference to the map that's still in `childTokens`.  The driversMtx is supposed to be guarding that so the need copy of the map is necessary.  Copying the `dex.Token` would add some extra safety, but I think consumers won't modify it.

File: server/market/balancer.go [hunk 5]
```diff
@@ -33,32 +39,88 @@ type MatchNegotiator interface {
...
+			feeFamily: make(map[uint32]*dex.Asset),
+			markets:   markets,
+		}
+		balancers[assetID] = bb
+
+		isToken, parentID := asset.IsToken(assetID)
+		if isToken {
+			parent, found := balancers[parentID]
+			if !found {
+				return fmt.Errorf("%s parent asset %d(%s)", ba.Symbol, parentID, dex.BipIDSymbol(parentID))
```
[the remark this answers is not in the document]
Comment (chappjc) [reply server/market/balancer.go:78]: Yeah, agree, or "requires parent asset"

File: server/market/balancer.go [hunk 6]
```diff
@@ -77,43 +139,123 @@ func (b *DEXBalancer) CheckBalance(acctAddr string, assetID uint32, qty, lots ui
...
+	}
+
+	qty += addPending(assetID)
+
+	// Add fee-family asset pending fees.
+	for famID := range backedAsset.feeFamily {
+		if q := addPending(famID); isToken && famID == feeID {
+			// Add the quantity of the base chain order reserves to our fee
+			// asset balance check.
+			feeQty = q
```
Comment (chappjc) [inline server/market/balancer.go:225]: Depending on the intent here (I'm not completely studied yet), but would this need to be either a `+=` or left as `=` but with a break?

File: server/asset/driver.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (buck54321) [reply server/asset/driver.go:16]: See also https://github.com/decred/dcrdex/blob/b3f438ad2cccb525e6b663bab9710e84b0c86b6a/dex/bip-id.go#L592-L594

File: server/market/balancer.go (hunk 1 above)
Comment (buck54321) [reply server/market/balancer.go:178]: `redeemFees` is logged separately. Do you mind if I leave it? I think it helps with clarity.

File: server/market/balancer.go (hunk 6 above)
Comment (buck54321) [reply server/market/balancer.go:225]: > Edit: oh, are you after the side effects of addPending regardless of famID?

Yes.

File: server/market/balancer.go (hunk 1 above)
Comment (martonp) [reply server/market/balancer.go:178]: Ah yeah np.

File: server/asset/driver.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (buck54321) [reply server/asset/driver.go:179]: > The driversMtx is supposed to be guarding that so the deep copy of the map is necessary

Or can we just get rid of the `driversMtx`? I'm don't understand its purpose.

[the remark this answers is not in the document]
Comment (chappjc) [reply server/asset/driver.go:172]: Not sure.  I guess it was put their out of uncertainty about possible concurrency in calls to package's `init` functions...

**Review by chappjc — APPROVED** [review]

**Pull request merged**